File: framework/src/play/src/main/scala/play/api/libs/Jsonp.scala
/*
 * Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
 */
package play.api.libs

import play.api.libs.json.JsValue
import play.api.http.{ ContentTypeOf, ContentTypes, Writeable }
import play.api.mvc.Codec

/**
 * JSONP helper.
 *
 * Example of use, provided the following route definition:
 * {{{
 *   GET  /my-service       Application.myService(callback: String)
 * }}}
 * The following action definition:
 * {{{
 *   def myService(callback: String) = Action {
 *     val json = ...
 *     Ok(Jsonp(callback, json))
 *   }
 * }}}
 * And the following request:
 * {{{
 *   GET /my-service?callback=foo
 * }}}
 * The response will have content type “text/javascript” and will look like the following:
 * {{{
 *   foo({...});
 * }}}
 *
 * Another example, showing how to serve either JSON or JSONP from the same action, according to the presence of
 * a “callback” parameter in the query string:
 * {{{
 *   def myService = Action { implicit request =>
 *     val json = ...
 *     request.queryString.get("callback").flatMap(_.headOption) match {
 *       case Some(callback) => Ok(Jsonp(callback, json))
 *       case None => Ok(json)
 *     }
 *   }
 * }}}
 */
case class Jsonp(padding: String, json: JsValue)

object Jsonp {

  implicit def contentTypeOf_Jsonp(implicit codec: Codec): ContentTypeOf[Jsonp] = {
    ContentTypeOf[Jsonp](Some(ContentTypes.JAVASCRIPT))
  }

  implicit def writeableOf_Jsonp(implicit codec: Codec): Writeable[Jsonp] = Writeable { jsonp =>
    codec.encode("%s(%s);".format(jsonp.padding, jsonp.json))
  }

}

File: framework/src/play-java/src/main/java/play/libs/Jsonp.java
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs;

import com.fasterxml.jackson.databind.JsonNode;
import play.mvc.Http.MimeTypes;
import play.twirl.api.Content;

/**
 * The JSONP Content renders a JavaScript call of a JSON object.<br>
 * Example of use, provided the following route definition:
 * <pre>
 *   GET  /my-service        Application.myService(callback: String)
 * </pre>
 * The following action definition:
 * <pre>
 *   public static Result myService(String callback) {
 *     JsonNode json = ...
 *     return ok(jsonp(callback, json));
 *   }
 * </pre>
 * And the following request:
 * <pre>
 *   GET  /my-service?callback=foo
 * </pre>
 * The response will have content type "application/javascript" and will look like the following:
 * <pre>
 *   foo({...});
 * </pre>
 */
public class Jsonp implements Content {

    public Jsonp(String padding, JsonNode json) {
        this.padding = padding;
        this.json = json;
    }

    @Override
    public String body() {
        return padding + "(" + Json.stringify(json) + ");";
    }

    @Override
    public String contentType() {
        return MimeTypes.JAVASCRIPT;
    }

    private final String padding;
    private final JsonNode json;

    /**
     * @param padding Name of the callback
     * @param json Json content
     * @return A JSONP Content using padding and json.
     */
    public static Jsonp jsonp(String padding, JsonNode json) {
        return new Jsonp(padding, json);
    }

}
